src/controllers/plugins/api/HTTP.cpp

```diff
@@ -0,0 +1,462 @@
+{
+    lua::StackGuard guard(L, -1);
+    auto ptr = HTTPRequest::getOrError(L, 1);
+    return ptr->set_payload(L);
+}
+
+int HTTPRequest::set_payload(lua_State *L)
+{
+    auto top = lua_gettop(L);
+    if (top != 1)
+    {
+        luaL_error(
+            L, "HTTPRequest:set_payload needs 1 argument (a string payload)");
+        return 0;
+    }
+
+    std::string temporary;
+    if (!lua::pop(L, &temporary))
+    {
+        luaL_error(L,
+                   "HTTPRequest:set_payload failed to get payload, expected a "
+                   "string");
+        return 0;
+    }
+    this->req_ =
+        std::move(this->req_).payload(QByteArray::fromStdString(temporary));
+    return 0;
+}
+
+int HTTPRequest::set_header_wrap(lua_State *L)
+{
+    lua::StackGuard guard(L, -1);
+    auto ptr = HTTPRequest::getOrError(L, 1);
+    return ptr->set_header(L);
+}
+
+int HTTPRequest::set_header(lua_State *L)
+{
+    auto top = lua_gettop(L);
+    if (top != 2)
+    {
+        luaL_error(L, "HTTPRequest:set_header needs 2 arguments (a header name "
+                      "and a value)");
+        return 0;
+    }
+
+    std::string value;
+    if (!lua::pop(L, &value))
+    {
+        luaL_error(L,
+                   "cannot get value (2nd argument of HTTPRequest:set_header)");
+        return 0;
+    }
+    std::string name;
+    if (!lua::pop(L, &name))
+    {
+        luaL_error(L,
+                   "cannot get name (1st argument of HTTPRequest:set_header)");
+        return 0;
+    }
+    this->req_ = std::move(this->req_)
+                     .header(QByteArray::fromStdString(name),
+                             QByteArray::fromStdString(value));
+    return 0;
+}
+
+int HTTPRequest::create(lua_State *L)
+{
+    lua::StackGuard guard(L, -1);
+    if (lua_gettop(L) != 2)
+    {
+        luaL_error(L, "HTTPRequest.create needs exactly 2 arguments (method "
+                      "and url)");
+        lua_pushnil(L);
+        return 1;
+    }
+    QString url;
+    if (!lua::pop(L, &url))
+    {
+        luaL_error(L, "cannot get url (2nd argument of HTTPRequest.create, "
+                      "expected a string)");
+        lua_pushnil(L);
+        return 1;
+    }
+    auto parsedurl = QUrl(url);
+    if (!parsedurl.isValid())
+    {
+        luaL_error(L, "cannot parse url (2nd argument of HTTPRequest.create, "
+                      "got invalid url in argument)");
+        lua_pushnil(L);
+        return 1;
+    }
+    NetworkRequestType method{};
+    if (!lua::pop(L, &method))
+    {
+        luaL_error(L, "cannot get method (1st argument of HTTPRequest.create, "
+                      "expected a string)");
+        lua_pushnil(L);
+        return 1;
+    }
+    auto *pl = getIApp()->getPlugins()->getPluginByStatePtr(L);
+    if (!pl->hasHTTPPermissionFor(parsedurl))
+    {
+        luaL_error(L, "Plugin does not have permission to send HTTP requests "
+                      "to this URL");
+        return 0;
```

Comment:
What makes this return 0 instead of 1 - is it because we don't push nil?
If so, why don't we push nil here?

Reply:
`luaL_error` is technically `no return`. I'll update this to return imminently after calling `luaL_error`